### PR TITLE
LPS-148049 Keep only thirty party dependencies in package.json

### PR DESCRIPTION
**Pull request by ling-alan-huang**
https://issues.liferay.com/browse/LPS-148049

### CODE REVIEW

**Comment by ling-alan-huang** [comment]
ci:test:sf

**Comment by ling-alan-huang** [comment]
Hey @natocesarrego 
Please see SF failures, should we also add `modules/apps/item-selector/item-selector-taglib` in exceptions list?

**Comment by natocesarrego** [comment]
Yes, we should, @ling-alan-huang, because it's part of the internal dependencies shared among the portal code, please see the complete list [here](https://github.com/liferay/liferay-portal/blob/7707483ca38025bd57fdc5ab828f82ffe1932c3f/modules/npmscripts.config.js#L32-L277). :slightly_smiling_face: 

I'll start to review the pull now.

Thanks

File: source-formatter.properties [hunk 1]
```diff
@@ -550,8 +550,7 @@
         modules/apps/frontend-,\
         modules/apps/site-initializer/site-initializer-raylife,\
         modules/apps/site-initializer/site-initializer-testray,\
-        modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal,\
-        modules/sdk
```
Comment (natocesarrego) [inline source-formatter.properties]: @ling-alan-huang, can we keep `modules/sdk` here for now?

AFAIK, this folder is not compiled during `ant all` and we may introduce issues if we not exclude it.

File: modules/apps/archived/youtube-web/package.json [hunk 2]
```diff
@@ -1,6 +1,5 @@
 {
 	"dependencies": {
-		"frontend-js-web": "*"
```
Comment (natocesarrego) [inline modules/apps/archived/youtube-web/package.json:3]: @ling-alan-huang, can we add the folder `modules/apps/archived/` to the exclusions of this SF rule in `source-formatter.properties`? The modules inside this directory usually are deprecated and aren't part of `ant all`.

**Review by natocesarrego — COMMENTED** [review]
@ling-alan-huang, good job! :smile: 

Please see my comments. The first 2 comments are actually suggestions, thus feel free to accept them or not. :slightly_smiling_face: 

Thanks

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/JSONPackageJSONRedundantDependenciesCheck.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+		int x = content.indexOf("imports: {");
+
+		if (x == -1) {
+			return _internalDependenciesNames;
+		}
+
+		x = x + 9;
+
+		int y = x;
```
[the remark this answers is not in the document]
Comment (izaera) [reply modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/JSONPackageJSONRedundantDependenciesCheck.java:110]: 9 = string length of `imports: ` I guess...

File: modules/dxp/apps/commerce/commerce-organization-web/package.json [hunk 4]
```diff
@@ -1,6 +1,5 @@
 {
 	"dependencies": {
-		"d3": "5.9.3"
```
[the remark this answers is not in the document]
Comment (izaera) [reply modules/dxp/apps/commerce/commerce-organization-web/package.json:3]: :thinking: Currently `d3-...` libs are being provided by `frontend-taglib-chart`:

image

In fact, the exclusions you have linked are for certain files inside those packages that don't need to be exported. Those exclusion were probably added to optimize the build.

Then, the answer to:

> Therefore, should we remove d3 and its variations (d3-*) from modules/npmscripts.config.js and restore it to the package.json of modules commerce-organization-web and commerce-shop-by-diagram-web?

Is no.

[the remark this answers is not in the document]
Comment (natocesarrego) [reply modules/dxp/apps/commerce/commerce-organization-web/package.json:3]: Thank you very much for your clarification, @izaera! :smile:

File: source-formatter.properties [hunk 5]
```diff
@@ -546,6 +546,13 @@
 
     source.check.JSONPackageJSONRedundantDependenciesCheck.enabled=true
 
+    source.check.JSONPackageJSONRedundantDependenciesCheck.excludedDirNames=\
+        modules/apps/frontend-,\
```
[the remark this answers is not in the document]
Comment (ling-alan-huang) [reply source-formatter.properties:550]: Thanks @natocesarrego and @izaera for helping on SF task.

We have the same view that SF will skip checking excluded path that they are in this property(`source.check.JSONPackageJSONRedundantDependenciesCheck.excludedDirNames`).

I still need to check with you about the exclusions. It's really hard for me to understand the relationship about the frontend dependencies.


> can we exclude only the frontend modules described on liferay-portal/modules/npmscripts.config.js

So SF should skip checking not only the modules that we added in that property, but also the frontend modues described in `npmscripts.config.js`. I am saying that correctlly?

If yes, I would properbly be confused.
If we skip checking the frontend modules described on `npmscripts.config.js`, why did you remove some dependencies from them. see your changes on [modules/apps/frontend-js/frontend-js-components-web/package.json](https://github.com/brianchandotcom/liferay-portal/pull/113825/files#diff-98e8f13a9c32a3c78511897b945e02f38d79cb0c87721c9b0e104fc700eae665L3-L10)

`frontend-js-components-web` is in [npmscripts.config.js](https://github.com/liferay/liferay-portal/blob/7707483ca38025bd57fdc5ab828f82ffe1932c3f/modules/npmscripts.config.js#L99)

[the remark this answers is not in the document]
Comment (natocesarrego) [reply source-formatter.properties:550]: > So SF should skip checking not only the modules that we added in that property, but also the frontend modues described in npmscripts.config.js. I am saying that correctlly?

Yes!

>If yes, I would properbly be confused.
If we skip checking the frontend modules described on npmscripts.config.js, why did you remove some dependencies from them. see your changes on ...

Great question! For sure, my pull can lead to confusion, but I'll try to clarify everything to you. :slightly_smiling_face: 

Ideally, a specific dependency should be placed in a `package.json` of a single module and then this dependency should be served to all other modules demanding it. For example, the `frontend-js-components-web` declares the dependency `@clayui/card`, but the module `frontend-taglib-clay` also declares this same dependency, so we've a redundancy. Therefore, only one of these modules should declare `@clayui/card` and then serve it to who is demanding it. After my pull, only the `frontend-taglib-clay` declares the dependency `@clayui/card` in `package.json` and serve it to `frontend-js-components-web`. The idea of my pull was remove all these redundancies even if they're happening among `frontend-*` modules.

In addition, it's relevant to highlight that only a subset of the `frontend-*` modules will be responsible for serve dependencies to other modules, but some of them won't need to serve any dependency, consequently, their `dependencies` in `package.json` will be empty, as [frontend-theme-classic-style-guide-sample-web](https://github.com/brianchandotcom/liferay-portal/pull/113825/files#diff-bdf08dc9883bd2daec3a66362e569e7e06c9d770bc3f97b146ce17c90ed4ef90), for example.

Is this clearer now?

If not, please let me know and I'll be glad to help. :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (ling-alan-huang) [reply source-formatter.properties:550]: Thanks @natocesarrego for the detailed informations. It's clearer now, I will resend the pull request.

**Comment by ling-alan-huang** [comment]
Resent https://github.com/natocesarrego/liferay-portal/pull/1177

**Pull request closed without merging**